protocol/auditlog.go

```diff
@@ -0,0 +1,180 @@
+// This module implements a CONIKS audit log that a CONIKS auditor
+// maintains.
+// An audit log is a mirror of many CONIKS key directories' STR history,
+// allowing CONIKS clients to audit the CONIKS directories.
+
+package protocol
+
+import (
+	"github.com/coniks-sys/coniks-go/crypto/sign"
+)
+
+type directoryHistory struct {
+	signKey   sign.PublicKey
+	snapshots map[uint64]*DirSTR
+	latestSTR *DirSTR
+}
+
+// A ConiksAuditLog maintains the histories
+// of all CONIKS directories known to a CONIKS auditor.
+// Each history includes the directory's public signing key
+// enabling the auditor to verify the corresponding signed
+// tree roots.
+type ConiksAuditLog struct {
+	histories map[string]*directoryHistory
```

Comment:
> can we use the hash of the key server's public signing key?

I've been thinking about this some more, and on second thought, I'm not sure we should index the STRs in an audit log by the hash of a server's signing key. If we do, we would break the continuity of a server's history when the server changes its signing key. This doesn't seem ideal since it might allow a server to equivocate using its two different signing keys. Can you please explain the purpose of this change?